src/afrolabs/components/kafka.clj

```diff
@@ -393,6 +393,32 @@
       (pre-process-messages [_ msgs]
         (map #(update % :value value-fn') msgs)))))
 
+;; We have this "wrapper" for a producer because the strategy system
+;; attempts to call -comp/halt on any strategies. If this is called on
+;; the producer itself, it is closed/halted BEFORE it has finished producing
+;; records that are in flight. By wrapping it like this, we are shielding it
+;; from that -comp/halt call, and instead integrant will call halt on the producer
+;; at the right time, depending on the dependency chain.
+(defstrategy ProduceConsumerResultsWithProducer
+  [producer]
+  (when-not (s/valid? ::kafka-producer producer)
+    (throw (ex-info "`ProduceConsumerResultsWithProducer` requires a producer component."
+                    {:producer-value producer})))
+  (reify
+    IConsumedResultsHandler
+    (handle-consumption-results [_ msgs]
+      (let [xs (s/conform ::producer-consumption-results-xs-spec msgs)]
+        (when (= ::s/invalid xs)
+          ;; horrible place to throw, but this is the best we've got.
+          (throw (ex-info "The producer can only handle the result of consume-messages, if the value is either a single message to be produced, or a collection of messages to be produced."
+                          {::explain-str (s/explain-str ::producer-consumption-results-xs-spec msgs)
+                           ::explain-data (s/explain-data ::producer-consumption-results-xs-spec msgs)})))
+        (let [[xs-type _] xs]
+          (produce! producer
+                    (condp = xs-type
+                      :single [msgs]
+                      :more   msgs)))))))
```

Comment:
We use this, so we can shield a producer from having its `component/halt` called when a consumer's strategies are halted. Back long ago, we made the decision to have producer's implement the consumer strategy `IConsumedResultsHandler` so they could be used in-place for a consumer's strategies collection. This has worked with remarkably little problems this far. 

In the situation of testing `latest-device-updates`, I discovered an exception where a producer was closed, before it had finished producing records, because it was halted in one consumer, while another was still using it. This strategy prevents that from happening, so integrant now cleans up the producer based on component-dependency order.